### PR TITLE
add map-root flag

### PR DESCRIPTION
**Pull request by miekg**
This maps the root user to another ID, meaning no pods will be run with
root perms. This is an alternative approach to the failed, run systemk
as a user.

Test the security context parsing and make it return an error as well.

Signed-off-by: Miek Gieben <miek@miek.nl>

### CODE REVIEW

**Comment by pires** [comment]
Have you considered moving to a configuration file instead of flags?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [virtual-kubelet/systemk] add m..." ]
>Have you considered moving to a configuration file instead of flags?

No, because we only have a few flags. And a config file is another thing that needs
provising and exist.

File: cmd/flag.go [hunk 1]
```diff
@@ -37,6 +37,7 @@ func installFlags(flags *pflag.FlagSet, c *provider.Opts) {
 	flags.StringVar(&c.ListenAddress, "addr", provider.DefaultListenAddr, "address to bind for serving requests from the Kubernetes API server")
 	flags.StringVar(&c.MetricsAddr, "metrics-addr", provider.DefaultMetricsAddr, "address to listen for metrics/stats requests")
 	flags.IntVar(&c.PodSyncWorkers, "pod-sync-workers", provider.DefaultPodSyncWorkers, `number of pod synchronization workers`)
+	flags.IntVar(&c.MapRoot, "map-root", 0, "map the root user to this UID")
```
Comment (pires) [inline cmd/flag.go:40]: The flag name is non-intuitive to me, so I suggest something like `override-root-uid`.

File: internal/provider/opts.go [hunk 2]
```diff
@@ -103,6 +103,9 @@ type Opts struct {
 	// UserMode is whether to use the user's systemd or the system's.
 	UserMode bool
 
+	// MapRoot maps the root user to this UID (defaults to 0).
+	MapRoot int
```
Comment (pires) [inline internal/provider/opts.go:107]: Given it's an int, we should ensure its validation in `SetDefaults`. Can you please add such validation?

File: cmd/flag.go (hunk 1 above)
Comment (miekg) [reply cmd/flag.go]: done

File: internal/provider/opts.go (hunk 2 above)
Comment (miekg) [reply internal/provider/opts.go]: to check for <0? You also want SetDefaults to return an error in that case?
Note user.LookupId with a negative number will fail

Comment (pires) [reply internal/provider/opts.go]: Either return an error or override to zero, but I guess an error is good to convey a misconfiguration issue to the user.

Comment (pires) [reply internal/provider/opts.go]: Ensuring the ID actually exists also makes sense to me if that's what you're proposing.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [virtual-kubelet/systemk] add m..." ]
>@pires commented on this pull request.
>
>━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
>
>In internal/provider/opts.go:
>
>> @@ -103,6 +103,9 @@ type Opts struct {
>        // UserMode is whether to use the user's systemd or the system's.
>        UserMode bool
>
>+       // MapRoot maps the root user to this UID (defaults to 0).
>+       MapRoot int
>
>Ensuring the ID actually exists also makes sense to me if that's what you're
>proposing.

that won't work, the user may be added later.

I'll add a < 0 check.

**Comment by miekg** [comment]
>Ensuring the ID actually exists also makes sense to me if that's what you're
>proposing.

added check

File: main.go [hunk 3]
```diff
@@ -72,6 +74,5 @@ func main() {
 		log.Fatal(err)
 	}
 
-	log.Info("system exited gracefully")
-
+	log.Info("systemk exited gracefully")
```
Comment (pires) [inline main.go:77]: Oops 😅

**Review by pires — APPROVED** [review]
LGTM

**Pull request merged**